File: backend/app/department_api.py
from __future__ import annotations

import re
import uuid
from datetime import date
from typing import Annotated, Literal, cast

from fastapi import APIRouter, Depends, Path, Request, Response, status
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncConnection

from app.auth.application_user import AuthorizationPrincipal
from app.auth.dependencies import (
    AdminSelectedBranch,
    AuthenticatedReadPrincipal,
    AuthenticatedWritePrincipal,
    VerifiedAccessToken,
    operation_not_permitted_error,
)
from app.http.errors import api_error, error_response_documentation, success_response_documentation
from app.http.idempotency import parse_idempotency_key
from app.http.idempotency_fingerprint import request_fingerprint
from app.http.schemas import CollectionResponse, DataResponse, Page
from app.http.validation import parse_pagination, parse_sort, validate_query_parameters
from app.models.identity import AppRole
from app.repositories.idempotency import IdempotencyRepository
from app.schemas.departments import (
    DepartmentCreateRequest,
    DepartmentDeleteRequest,
    DepartmentResponse,
    DepartmentUpdateRequest,
    StaffingRuleCreateRequest,
    StaffingRuleDeleteRequest,
    StaffingRuleResponse,
    StaffingRuleUpdateRequest,
)
from app.services.departments import (
    DepartmentListQuery,
    DepartmentService,
    StaffingRuleListQuery,
    normalize_department,
    normalize_search,
)
from app.services.execution import AuthorizedServiceExecutor
from app.services.idempotency import IdempotencyCommand, IdempotencyCoordinator, IdempotentResponse

router = APIRouter(prefix="/api/v1", tags=["departments"])
CANONICAL_UUID_PATTERN = r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
_CANONICAL_UUID = re.compile(CANONICAL_UUID_PATTERN)
_DATE = re.compile(r"^\d{4}-\d{2}-\d{2}$")

DEPARTMENT_ERRORS = error_response_documentation(
    "invalid_request",
    "invalid_access_token",
    "application_account_unavailable",
    "operation_not_permitted",
    "origin_not_allowed",
    "resource_not_found",
    "method_not_allowed",
    "not_acceptable",
    "request_too_large",
    "unsupported_media_type",
    "validation_failed",
    "invalid_branch",
    "invalid_cursor",
    "state_conflict",
    "department_conflict",
    "staffing_rule_conflict",
    "idempotency_key_required",
    "invalid_idempotency_key",
    "idempotency_conflict",
    "idempotency_in_progress",
    "rate_limit_exceeded",
    "application_account_lookup_unavailable",
    "request_timeout",
    "internal_error",
)


def _service(request: Request, connection: AsyncConnection) -> DepartmentService:
    factory = getattr(request.app.state, "department_service_factory", None)
    if factory is not None:
        return cast(DepartmentService, factory(connection))
    return DepartmentService(connection, request.app.state.department_cursor_codec)


def _idempotency(request: Request, connection: AsyncConnection) -> IdempotencyCoordinator:
    factory = getattr(request.app.state, "idempotency_coordinator_factory", None)
    if factory is not None:
        return cast(IdempotencyCoordinator, factory(connection))
    return IdempotencyCoordinator(IdempotencyRepository(connection))


def _require_admin(principal: AuthorizationPrincipal) -> None:
    if principal.role is not AppRole.ADMIN:
        raise operation_not_permitted_error()


def _nullable_uuid(value: str | None) -> uuid.UUID | Literal["null"] | None:
    if value is None or value == "null":
        return value
    if not _CANONICAL_UUID.fullmatch(value):
        raise ValueError("invalid UUID")
    return uuid.UUID(value)


def _department_query(request: Request) -> DepartmentListQuery:
    validate_query_parameters(
        request, allowed={"limit", "cursor", "search", "parentId", "headEmployeeId", "sort"}
    )
    try:
        pagination = parse_pagination(
            limit=request.query_params.get("limit"), cursor=request.query_params.get("cursor")
        )
        return DepartmentListQuery(
            limit=pagination.limit,
            search=normalize_search(request.query_params.get("search")),
            parent_id=_nullable_uuid(request.query_params.get("parentId")),
            head_employee_id=_nullable_uuid(request.query_params.get("headEmployeeId")),
            sort=parse_sort(
                request.query_params.get("sort"),
                allowed_fields={"sortOrder", "name", "createdAt"},
                default=("sortOrder", "name"),
            ),
            cursor=pagination.cursor,
        )
    except ValueError:
        raise api_error("validation_failed") from None


def _staffing_query(request: Request) -> StaffingRuleListQuery:
    validate_query_parameters(
        request,
        allowed={"limit", "cursor", "department", "shiftCategory", "effectiveOn", "sort"},
    )
    try:
        pagination = parse_pagination(
            limit=request.query_params.get("limit"), cursor=request.query_params.get("cursor")
        )
        effective_on_raw = request.query_params.get("effectiveOn")
        effective_on = None
        if effective_on_raw is not None:
            if _DATE.fullmatch(effective_on_raw) is None:
                raise ValueError("invalid date")
            effective_on = date.fromisoformat(effective_on_raw)
        category = request.query_params.get("shiftCategory")
        if category is not None and category not in {"morning", "afternoon", "night", "flexible"}:
            raise ValueError("invalid category")
        return StaffingRuleListQuery(
            limit=pagination.limit,
            department=normalize_department(request.query_params.get("department")),
            shift_category=category,
            effective_on=effective_on,
            sort=parse_sort(
                request.query_params.get("sort"),
                allowed_fields={"department", "shiftCategory", "effectiveFrom", "effectiveTo"},
                default=("department", "shiftCategory"),
            ),
            cursor=pagination.cursor,
        )
    except ValueError:
        raise api_error("validation_failed") from None


DepartmentQuery = Annotated[DepartmentListQuery, Depends(_department_query)]
StaffingQuery = Annotated[StaffingRuleListQuery, Depends(_staffing_query)]


def _path_id(value: str) -> uuid.UUID:
    if not _CANONICAL_UUID.fullmatch(value):
        raise api_error("validation_failed")
    return uuid.UUID(value)


@router.get(
    "/departments",
    response_model=CollectionResponse[DepartmentResponse],
    operation_id="list_departments",
    responses={
        **success_response_documentation(
            200, "Selected branch departments", cache_control="no-store"
        ),
        **DEPARTMENT_ERRORS,
    },
)
async def list_departments(
    request: Request,
    claims: VerifiedAccessToken,
    principal: AuthenticatedReadPrincipal,
    selected_branch_id: AdminSelectedBranch,
    query: DepartmentQuery,
) -> CollectionResponse[DepartmentResponse]:
    _require_admin(principal)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> tuple[list[DepartmentResponse], str | None]:
        return await _service(request, connection).list_departments(
            principal, selected_branch_id, query
        )

    items, cursor = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return CollectionResponse(
        data=items,
        page=Page(limit=query.limit, next_cursor=cursor, has_more=cursor is not None),
    )


@router.post(
    "/departments",
    status_code=status.HTTP_201_CREATED,
    response_model=DataResponse[DepartmentResponse],
    operation_id="create_department",
    responses={
        **success_response_documentation(201, "Created department", cache_control="no-store"),
        **DEPARTMENT_ERRORS,
    },
)
async def create_department(
    request: Request,
    body: DepartmentCreateRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
) -> Response:
    _require_admin(principal)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> DepartmentResponse:
        return await _service(request, connection).create_department(
            principal, selected_branch_id, body
        )

    item = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return JSONResponse(
        status_code=201,
        content=DataResponse(data=item).model_dump(mode="json", by_alias=True),
        headers={"Cache-Control": "no-store", "Location": f"/api/v1/departments/{item.id}"},
    )


@router.patch(
    "/departments/{department_id}",
    response_model=DataResponse[DepartmentResponse],
    operation_id="update_department",
    responses={
        **success_response_documentation(200, "Updated department", cache_control="no-store"),
        **DEPARTMENT_ERRORS,
    },
)
async def update_department(
    request: Request,
    body: DepartmentUpdateRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
    department_id: Annotated[str, Path(pattern=CANONICAL_UUID_PATTERN)],
) -> Response:
    _require_admin(principal)
    parsed_id = _path_id(department_id)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor
    name_supplied = "name" in body.model_fields_set
    key = parse_idempotency_key(request, required=name_supplied)

    async def operation(connection: AsyncConnection) -> IdempotentResponse:
        service = _service(request, connection)

        async def mutate() -> IdempotentResponse:
            item = await service.update_department(principal, selected_branch_id, parsed_id, body)
            return IdempotentResponse(
                status=200,
                body=DataResponse(data=item).model_dump(mode="json", by_alias=True),
                location=None,
                resource_kind="department",
                resource_id=item.id,
            )

        if key is None:
            return await mutate()
        route_parameters: dict[str, object] = {"departmentId": str(parsed_id)}
        command = IdempotencyCommand(
            key=key,
            operation_id="update_department",
            method="PATCH",
            route_parameters=route_parameters,
            fingerprint=request_fingerprint(
                operation_id="update_department",
                method="PATCH",
                route_parameters=route_parameters,
                effective_query_parameters={},
                body=body.model_dump(mode="json", by_alias=True),
            ),
            branch_id=selected_branch_id,
        )
        return await _idempotency(request, connection).execute(
            principal=principal,
            command=command,
            authorize_replay=lambda kind, resource_id: service.authorize_department_replay(
                principal, selected_branch_id, kind, resource_id
            ),
            mutation=mutate,
        )

    outcome = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    headers = {"Cache-Control": "no-store"}
    if outcome.replayed:
        headers["Idempotency-Replayed"] = "true"
    return JSONResponse(status_code=outcome.status, content=outcome.body, headers=headers)


@router.delete(
    "/departments/{department_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    operation_id="delete_department",
    responses={
        **success_response_documentation(204, "Deleted department"),
        **DEPARTMENT_ERRORS,
    },
)
async def delete_department(
    request: Request,
    body: DepartmentDeleteRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
    department_id: Annotated[str, Path(pattern=CANONICAL_UUID_PATTERN)],
) -> Response:
    _require_admin(principal)
    parsed_id = _path_id(department_id)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> None:
        await _service(request, connection).delete_department(
            principal, selected_branch_id, parsed_id, body
        )

    await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return Response(status_code=204, headers={"Cache-Control": "no-store"})


@router.get(
    "/department-staffing-rules",
    response_model=CollectionResponse[StaffingRuleResponse],
    operation_id="list_department_staffing_rules",
    responses={
        **success_response_documentation(
            200, "Selected branch staffing rules", cache_control="no-store"
        ),
        **DEPARTMENT_ERRORS,
    },
)
async def list_staffing_rules(
    request: Request,
    claims: VerifiedAccessToken,
    principal: AuthenticatedReadPrincipal,
    selected_branch_id: AdminSelectedBranch,
    query: StaffingQuery,
) -> CollectionResponse[StaffingRuleResponse]:
    _require_admin(principal)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(
        connection: AsyncConnection,
    ) -> tuple[list[StaffingRuleResponse], str | None]:
        return await _service(request, connection).list_staffing_rules(
            principal, selected_branch_id, query
        )

    items, cursor = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return CollectionResponse(
        data=items,
        page=Page(limit=query.limit, next_cursor=cursor, has_more=cursor is not None),
    )


@router.post(
    "/department-staffing-rules",
    status_code=status.HTTP_201_CREATED,
    response_model=DataResponse[StaffingRuleResponse],
    operation_id="create_department_staffing_rule",
    responses={
        **success_response_documentation(201, "Created staffing rule", cache_control="no-store"),
        **DEPARTMENT_ERRORS,
    },
)
async def create_staffing_rule(
    request: Request,
    body: StaffingRuleCreateRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
) -> Response:
    _require_admin(principal)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> StaffingRuleResponse:
        return await _service(request, connection).create_staffing_rule(
            principal, selected_branch_id, body
        )

    item = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return JSONResponse(
        status_code=201,
        content=DataResponse(data=item).model_dump(mode="json", by_alias=True),
        headers={
            "Cache-Control": "no-store",
            "Location": f"/api/v1/department-staffing-rules/{item.id}",
        },
    )


@router.patch(
    "/department-staffing-rules/{rule_id}",
    response_model=DataResponse[StaffingRuleResponse],
    operation_id="update_department_staffing_rule",
    responses={
        **success_response_documentation(200, "Updated staffing rule", cache_control="no-store"),
        **DEPARTMENT_ERRORS,
    },
)
async def update_staffing_rule(
    request: Request,
    body: StaffingRuleUpdateRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
    rule_id: Annotated[str, Path(pattern=CANONICAL_UUID_PATTERN)],
) -> DataResponse[StaffingRuleResponse]:
    _require_admin(principal)
    parsed_id = _path_id(rule_id)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> StaffingRuleResponse:
        return await _service(request, connection).update_staffing_rule(
            principal, selected_branch_id, parsed_id, body
        )

    item = await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return DataResponse(data=item)


@router.delete(
    "/department-staffing-rules/{rule_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    operation_id="delete_department_staffing_rule",
    responses={
        **success_response_documentation(204, "Deleted staffing rule"),
        **DEPARTMENT_ERRORS,
    },
)
async def delete_staffing_rule(
    request: Request,
    body: StaffingRuleDeleteRequest,
    claims: VerifiedAccessToken,
    principal: AuthenticatedWritePrincipal,
    selected_branch_id: AdminSelectedBranch,
    rule_id: Annotated[str, Path(pattern=CANONICAL_UUID_PATTERN)],
) -> Response:
    _require_admin(principal)
    parsed_id = _path_id(rule_id)
    executor: AuthorizedServiceExecutor = request.app.state.authorized_service_executor

    async def operation(connection: AsyncConnection) -> None:
        await _service(request, connection).delete_staffing_rule(
            principal, selected_branch_id, parsed_id, body
        )

    await executor.execute(
        claims=claims,
        principal=principal,
        operation=operation,
        selected_admin_branch_id=selected_branch_id,
    )
    return Response(status_code=204, headers={"Cache-Control": "no-store"})

File: src/departmentApi.js
const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/
const uuid4Pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/
const instantPattern = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/
const datePattern = /^\d{4}-\d{2}-\d{2}$/
const categories = new Set(['morning', 'afternoon', 'night', 'flexible'])

function isRecord(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
}

function hasExactKeys(value, expected) {
  if (!isRecord(value)) return false
  const actual = Object.keys(value).sort()
  const required = [...expected].sort()
  return actual.length === required.length
    && actual.every((key, index) => key === required[index])
}

function invalidResponse() {
  return new Error('Invalid department response')
}

function parseDepartment(value) {
  if (
    !hasExactKeys(value, [
      'id', 'name', 'parentId', 'headEmployeeId', 'color', 'description', 'sortOrder', 'createdAt',
    ])
    || !uuidPattern.test(value.id)
    || typeof value.name !== 'string'
    || value.parentId !== null && !uuidPattern.test(value.parentId)
    || value.headEmployeeId !== null && !uuidPattern.test(value.headEmployeeId)
    || typeof value.color !== 'string'
    || typeof value.description !== 'string'
    || !Number.isInteger(value.sortOrder)
    || !instantPattern.test(value.createdAt)
  ) throw invalidResponse()
  return Object.freeze({ ...value })
}

function parseStaffingRule(value) {
  if (
    !hasExactKeys(value, [
      'id', 'department', 'shiftCategory', 'minStaff', 'effectiveFrom', 'effectiveTo',
    ])
    || !uuidPattern.test(value.id)
    || typeof value.department !== 'string'
    || !categories.has(value.shiftCategory)
    || !Number.isInteger(value.minStaff)
    || value.minStaff < 0
    || value.effectiveFrom !== null && !datePattern.test(value.effectiveFrom)
    || value.effectiveTo !== null && !datePattern.test(value.effectiveTo)
    || value.effectiveFrom !== null && value.effectiveTo !== null
      && value.effectiveTo < value.effectiveFrom
  ) throw invalidResponse()
  return Object.freeze({ ...value })
}

function parsePage(value) {
  if (
    !hasExactKeys(value, ['limit', 'nextCursor', 'hasMore'])
    || !Number.isInteger(value.limit)
    || value.limit < 1
    || value.limit > 100
    || value.nextCursor !== null && typeof value.nextCursor !== 'string'
    || typeof value.hasMore !== 'boolean'
    || value.hasMore !== (value.nextCursor !== null)
  ) throw invalidResponse()
  return Object.freeze({ ...value })
}

function collection(response, parser) {
  if (!Array.isArray(response.data)) throw invalidResponse()
  return Object.freeze({
    data: Object.freeze(response.data.map(parser)),
    page: parsePage(response.page),
  })
}

function queryString(options, allowed) {
  const parameters = new URLSearchParams()
  for (const [key, value] of Object.entries(options)) {
    if (value === undefined || key === 'signal') continue
    if (!allowed.has(key)) throw new TypeError('Invalid department query')
    if (key === 'limit' && (!Number.isInteger(value) || value < 1 || value > 100)) {
      throw new TypeError('Invalid department query')
    }
    if (['search', 'department'].includes(key) && (
      typeof value !== 'string' || !value.trim() || value.trim().length > (key === 'search' ? 100 : 200)
    )) throw new TypeError('Invalid department query')
    if (['parentId', 'headEmployeeId'].includes(key) && value !== 'null' && !uuidPattern.test(value)) {
      throw new TypeError('Invalid department query')
    }
    if (key === 'shiftCategory' && !categories.has(value)) throw new TypeError('Invalid staffing query')
    if (key === 'effectiveOn' && !datePattern.test(value)) throw new TypeError('Invalid staffing query')
    if (['cursor', 'sort'].includes(key) && (typeof value !== 'string' || !value)) {
      throw new TypeError('Invalid department query')
    }
    parameters.set(key, ['search', 'department'].includes(key) ? value.normalize('NFC').trim() : String(value))
  }
  const encoded = parameters.toString()
  return encoded ? `?${encoded}` : ''
}

async function readAll(readPage, options) {
  const rows = []
  const ids = new Set()
  const cursors = new Set()
  let cursor
  do {
    const page = await readPage({ ...options, cursor, limit: 100 })
    for (const row of page.data) {
      if (ids.has(row.id)) throw invalidResponse()
      ids.add(row.id)
      rows.push(row)
    }
    cursor = page.page.nextCursor ?? undefined
    if (cursor !== undefined) {
      if (cursors.has(cursor)) throw invalidResponse()
      cursors.add(cursor)
    }
  } while (cursor !== undefined)
  return Object.freeze(rows)
}

function requireBranch(branchId) {
  if (!uuidPattern.test(branchId)) throw new TypeError('Invalid branch ID')
}

function requestOptions(branchId, options = {}) {
  requireBranch(branchId)
  return { access: 'protected', headers: { 'X-Workloop-Branch-ID': branchId }, ...options }
}

const departmentQueries = new Set(['limit', 'cursor', 'search', 'parentId', 'headEmployeeId', 'sort'])
const staffingQueries = new Set(['limit', 'cursor', 'department', 'shiftCategory', 'effectiveOn', 'sort'])

export async function readDepartments(authentication, branchId, options = {}) {
  const { signal, ...query } = options
  const response = await authentication.request(
    `/api/v1/departments${queryString(query, departmentQueries)}`,
    requestOptions(branchId, { signal }),
  )
  return collection(response, parseDepartment)
}

export function readAllDepartments(authentication, branchId, options = {}) {
  return readAll(
    (page) => readDepartments(authentication, branchId, page),
    options,
  )
}

export async function readStaffingRules(authentication, branchId, options = {}) {
  const { signal, ...query } = options
  const response = await authentication.request(
    `/api/v1/department-staffing-rules${queryString(query, staffingQueries)}`,
    requestOptions(branchId, { signal }),
  )
  return collection(response, parseStaffingRule)
}

export function readAllStaffingRules(authentication, branchId, options = {}) {
  return readAll(
    (page) => readStaffingRules(authentication, branchId, page),
    options,
  )
}

function exactMutation(value, allowed, required) {
  if (
    !isRecord(value)
    || Object.keys(value).some((key) => !allowed.includes(key))
    || required.some((key) => !Object.hasOwn(value, key))
  ) throw new TypeError('Invalid department mutation')
  return value
}

const departmentFields = ['name', 'parentId', 'headEmployeeId', 'color', 'description', 'sortOrder']
const staffingFields = ['department', 'shiftCategory', 'minStaff', 'effectiveFrom', 'effectiveTo']

function validateDepartmentMutation(body) {
  if (Object.hasOwn(body, 'name') && (
    typeof body.name !== 'string' || !body.name.trim() || body.name.length > 200
  )) throw new TypeError('Invalid department mutation')
  for (const key of ['parentId', 'headEmployeeId']) {
    if (Object.hasOwn(body, key) && body[key] !== null && !uuidPattern.test(body[key])) {
      throw new TypeError('Invalid department mutation')
    }
  }
  if (Object.hasOwn(body, 'color') && (
    typeof body.color !== 'string' || body.color.length > 100
  )) throw new TypeError('Invalid department mutation')
  if (Object.hasOwn(body, 'description') && (
    typeof body.description !== 'string' || body.description.length > 2000
  )) throw new TypeError('Invalid department mutation')
  if (Object.hasOwn(body, 'sortOrder') && !Number.isInteger(body.sortOrder)) {
    throw new TypeError('Invalid department mutation')
  }
}

function validateStaffingMutation(body) {
  if (Object.hasOwn(body, 'department') && (
    typeof body.department !== 'string' || !body.department.trim() || body.department.length > 200
  )) throw new TypeError('Invalid staffing mutation')
  if (Object.hasOwn(body, 'shiftCategory') && !categories.has(body.shiftCategory)) {
    throw new TypeError('Invalid staffing mutation')
  }
  if (Object.hasOwn(body, 'minStaff') && (
    !Number.isInteger(body.minStaff) || body.minStaff < 0
  )) throw new TypeError('Invalid staffing mutation')
  for (const key of ['effectiveFrom', 'effectiveTo']) {
    if (Object.hasOwn(body, key) && body[key] !== null && !datePattern.test(body[key])) {
      throw new TypeError('Invalid staffing mutation')
    }
  }
  const start = body.effectiveFrom
  const end = body.effectiveTo
  if (start !== undefined && end !== undefined && start !== null && end !== null && end < start) {
    throw new TypeError('Invalid staffing mutation')
  }
}

function validateDepartmentExpected(value) {
  if (!hasExactKeys(value, departmentFields)) throw new TypeError('Invalid department snapshot')
  validateDepartmentMutation(value)
}

function validateStaffingExpected(value) {
  if (!hasExactKeys(value, staffingFields)) throw new TypeError('Invalid staffing snapshot')
  validateStaffingMutation(value)
}

export function departmentSnapshot(value) {
  const parsed = parseDepartment(value)
  return Object.freeze(Object.fromEntries(departmentFields.map((key) => [key, parsed[key]])))
}

export function staffingRuleSnapshot(value) {
  const parsed = parseStaffingRule(value)
  return Object.freeze(Object.fromEntries(staffingFields.map((key) => [key, parsed[key]])))
}

export async function createDepartment(authentication, branchId, values, { signal } = {}) {
  const body = exactMutation(values, departmentFields, ['name'])
  validateDepartmentMutation(body)
  const response = await authentication.request('/api/v1/departments', requestOptions(branchId, {
    method: 'POST', json: body, signal,
  }))
  const data = parseDepartment(response.data)
  if (response.status !== 201 || response.location !== `/api/v1/departments/${data.id}`) {
    throw invalidResponse()
  }
  return data
}

export async function updateDepartment(
  authentication,
  branchId,
  departmentId,
  changes,
  { idempotencyKey, signal } = {},
) {
  if (!uuidPattern.test(departmentId)) throw new TypeError('Invalid department ID')
  const body = exactMutation(changes, [...departmentFields, 'expected'], ['expected'])
  if (Object.keys(body).length === 1) throw new TypeError('Invalid department update')
  validateDepartmentMutation(body)
  validateDepartmentExpected(body.expected)
  const headers = { 'X-Workloop-Branch-ID': branchId }
  if (Object.hasOwn(body, 'name')) {
    if (!uuid4Pattern.test(idempotencyKey)) throw new TypeError('Invalid idempotency key')
    headers['Idempotency-Key'] = idempotencyKey
  }
  const response = await authentication.request(`/api/v1/departments/${departmentId}`, {
    access: 'protected', method: 'PATCH', headers, json: body, signal,
  })
  const data = parseDepartment(response.data)
  if (data.id !== departmentId) throw invalidResponse()
  return Object.freeze({ data, replayed: response.replayed })
}

export async function deleteDepartment(authentication, branchId, value, { signal } = {}) {
  const department = parseDepartment(value)
  const response = await authentication.request(`/api/v1/departments/${department.id}`, requestOptions(branchId, {
    method: 'DELETE', json: { expected: departmentSnapshot(department) }, signal,
  }))
  if (response.status !== 204 || response.data !== null) throw invalidResponse()
}

export async function createStaffingRule(authentication, branchId, values, { signal } = {}) {
  const body = exactMutation(values, staffingFields, staffingFields)
  validateStaffingMutation(body)
  const response = await authentication.request('/api/v1/department-staffing-rules', requestOptions(branchId, {
    method: 'POST', json: body, signal,
  }))
  const data = parseStaffingRule(response.data)
  if (response.status !== 201 || response.location !== `/api/v1/department-staffing-rules/${data.id}`) {
    throw invalidResponse()
  }
  return data
}

export async function updateStaffingRule(authentication, branchId, value, changes, { signal } = {}) {
  const rule = parseStaffingRule(value)
  const body = exactMutation(changes, [...staffingFields, 'expected'], ['expected'])
  if (Object.keys(body).length === 1) throw new TypeError('Invalid staffing update')
  validateStaffingMutation(body)
  validateStaffingExpected(body.expected)
  validateStaffingMutation({ ...body.expected, ...body })
  const { data } = await authentication.request(`/api/v1/department-staffing-rules/${rule.id}`, requestOptions(branchId, {
    method: 'PATCH', json: body, signal,
  }))
  const parsed = parseStaffingRule(data)
  if (parsed.id !== rule.id) throw invalidResponse()
  return parsed
}

export async function deleteStaffingRule(authentication, branchId, value, { signal } = {}) {
  const rule = parseStaffingRule(value)
  const response = await authentication.request(`/api/v1/department-staffing-rules/${rule.id}`, requestOptions(branchId, {
    method: 'DELETE', json: { expected: staffingRuleSnapshot(rule) }, signal,
  }))
  if (response.status !== 204 || response.data !== null) throw invalidResponse()
}
